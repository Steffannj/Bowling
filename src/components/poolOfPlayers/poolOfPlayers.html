<template>
  <require from="./poolOfPlayers.scss"></require>
  <form class="h-100" role="form" submit.delegate="startDraw()">
    <div class="form-group player-name">
      <label>Player name</label>
      <input type="text" value.bind="playerName" class="form-control"
        placeholder="Enter player name" disabled.bind="isBusy">
        <button class="btn btn-success" click.delegate="add(playerName)"  disabled.bind="isBusy">Add</button>
    </div>
    <div class="overflow-auto h-75">
      <div class="form-group players-input" repeat.for="player of players">
        <label>${player.name}</label>
      </div>
    </div>
    <button type="submit" class="btn btn-success w-100 mt-1 mb-1" disabled.bind="!canDraw || isBusy"
      show.bind="!drawCompleted">Start
      draw</button>
    <button class="btn btn-danger w-100 mt-1 mb-1" click.delegate="reset()" show.bind="drawCompleted">Reset
      draw</button>
  </form>
</template>
