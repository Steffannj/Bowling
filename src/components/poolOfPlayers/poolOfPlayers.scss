.player-name {
  display: flex;
  align-items: center;
  margin-top: 10px;

  label {
    flex: 1;
  }

  input {
    flex: 3;
  }
}
