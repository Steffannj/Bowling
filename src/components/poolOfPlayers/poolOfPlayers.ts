import { EventAggregator, Subscription } from 'aurelia-event-aggregator';
import { inject, observable } from 'aurelia-framework';
import { NewInstance } from 'aurelia-framework';
import { Player } from '../player/player';
import { ValidationRules, ValidationController } from 'aurelia-validation';
import * as Toastr from 'toastr';

@inject(EventAggregator)
export class PoolOfPlayers {
  playerName: string;
  @observable players: Array<Player>;
  @observable size: number;
  ea: EventAggregator;
  isBusy: boolean;
  canDraw: boolean;
  drawCompleted: boolean;
  drawEndedSubscription: Subscription;

  constructor(ea: EventAggregator) {
    this.players = [];
    this.ea = ea;
    this.size = 0;
  }

  attached() {
    this.drawEndedSubscription = this.ea.subscribe("drawCompleted", res => setTimeout(() => {
      this.drawCompleted = res
    }, 2500));
  }

  detached() {
    this.drawEndedSubscription.dispose();
  }

  sizeChanged(n, o) {
    if (n > 0 && n % 2 == 0)
      this.canDraw = true;
    else
      this.canDraw = false;
  }

  playersChanged(n, o) {
    console.log("asdasd");
  }

  add(playerName: string) {
    if (playerName) {
      this.players.push(new Player(playerName));
      this.size++;
      this.playerName = '';
      this.ea.publish('addPlayer', this.players);
    } else {
      Toastr.error("Player name can not be blank.");
    }
  }

  startDraw() {
    if (!this.canDraw)
      return;
    this.isBusy = true;
    this.ea.publish('startDraw', true);
  }

  reset() {
    this.players = [];
    this.size = 0;
    this.isBusy = false;
    this.drawCompleted = false;
    this.ea.publish('drawReseted', true);
  }
}
