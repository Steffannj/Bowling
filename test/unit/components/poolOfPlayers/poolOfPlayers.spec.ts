import { EventAggregator } from 'aurelia-event-aggregator';
import { PoolOfPlayers } from '../../../../src/components/poolOfPlayers/poolOfPlayers';

describe('The Pool of players module', () => {
  let sut: PoolOfPlayers;
  let ea: EventAggregator;

  beforeEach(() => {
    ea = new EventAggregator;
    sut = new PoolOfPlayers(ea);
  });

  it('should be defined', () => {
    expect(sut).toBeDefined();
  });

  it('should add player in players array if add is called', () => {
    sut.add('Stefan');
    expect(sut.players).toMatchSnapshot(
      [{ name: 'Stefan' }]
    );
  });

  it('should have size of 4 if 4 players are added', () => {
    sut.add('Stefan');
    sut.add('Zvjezdan');
    sut.add('Aner');
    sut.add('Mladen');
    expect(sut.size).toEqual(4);
  });

  it('should not add player if empty string is passed', () => {
    sut.add('');
    expect(sut.size).toEqual(0);
  });

  it('should set canDraw to true if num of players is even', () => {
    sut.add('Stefan');
    sut.add('Zvjezdan');
    expect(sut.canDraw).toBeTruthy();
  });

  it('should set canDraw to false if num of players is odd', () => {
    sut.add('Stefan');
    expect(sut.canDraw).toBeFalsy();
  });

  it('should clear array of players if reset is called', () => {
    sut.add('Stefan');
    sut.add('Zvjezdan');
    sut.reset();
    expect(sut.players.length).toEqual(0);
  });
  
  it('should set isBusy to false if reset is called', () => {
    sut.add('Stefan');
    sut.add('Zvjezdan');
    sut.startDraw();
    expect(sut.isBusy).toBeTruthy();
    sut.reset();
    expect(sut.isBusy).toBeFalsy();
  });

})
