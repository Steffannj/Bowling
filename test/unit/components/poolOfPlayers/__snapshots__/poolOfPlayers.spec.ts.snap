// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Pool of players module should add player in players array if add is called 1`] = `
Object {
  "0": Player {
    "drawed": false,
    "name": "Stefan",
  },
}
`;
